dirty_cat/_deduplicate.py

```diff
@@ -0,0 +1,175 @@
+    categories in data, performing hierarchical clustering on this distance
+    matrix, and choosing the most frequent element in each cluster as the
+    'correct' spelling. This method works best if the true number of
+    categories is significantly smaller than the number of observed spellings.
+
+    Parameters
+    ----------
+    data : Sequence[str]
+        The data to be deduplicated.
+    n_clusters : Optional[int], optional
+        number of clusters to use for hierarchical clustering, if None use the
+        number of clusters that lead to the lowest silhouette score,
+        by default None
+    ngram_range : Tuple[int, int], optional
+        range to use for computing n-gram distance, by default (2, 4)
+    analyzer : str, optional
+        `CountVectorizer` analyzer for computing n-grams, by default "char_wb"
+    method : str, optional
+        linkage method to use for merging clusters, by default "average"
+    return_translation_table : bool, optional
+        Whether to return a translation table with original words as indices
+        and deduplicated words as values, by default False
+
+    Returns
+    -------
+    Union[List[str], Tuple[List[str], pd.Series]]
+       Either the deduplicated data (if return_translation_table=False) or a
+       tuple of the deduplicated data and a translation table that maps
+       original data to deduplicated data (if return_translation_table=True).
+    """
+    unique_words, counts = np.unique(data, return_counts=True)
+    distance_mat = compute_ngram_distance(
+        unique_words, ngram_range=ngram_range, analyzer=analyzer
+    )
+
+    Z = linkage(distance_mat, method=method, optimal_ordering=True)
+    if n_clusters is None:
+        n_clusters = guess_clusters(Z, distance_mat)
+    clusters = fcluster(Z, n_clusters, criterion="maxclust")
+
+    translation_table = create_spelling_correction(unique_words, counts, clusters)
+    unrolled_corrections = translation_table[data]
+    if return_translation_table:
+        return unrolled_corrections, translation_table
+    else:
+        return unrolled_corrections
+
+
+def guess_clusters(Z: np.ndarray, distance_mat: np.ndarray) -> int:
+    """Finds the number of clusters that maximize the silhouette score
+    when clustering `distance_mat`.
+
+    Parameters
+    ----------
+    Z : np.ndarray
+        hierarchical linkage matrix, specifies which clusters to merge.
+    distance_mat : np.ndarray
+        distance matrix either in square or condensed form.
+
+    Returns
+    -------
+    int
+        number of clusters that maximize the silhouette score.
+    """
+    max_clusters = distance_mat.shape[0]
+    n_clusters = np.arange(2, max_clusters)
+    # silhouette score needs a redundant distance matrix
+    redundant_dist = squareform(distance_mat)
+    silhouette_scores = []
+    for n_clust in n_clusters:
+        labels = fcluster(Z, n_clust, criterion="maxclust")
```

Comment:
Why did you choose criterion `maxclust`?

Reply:
More out of ease-of-use than performance considerations.
I found it easier for me (or other users) to "guess" the number of unique categories that is present in the data, than have an intuition about minimum cluster distance etc.